File: String/3LongestSubstringWithoutRepeatingCharacters.cpp
/**
 Solution:
 Define a mapping of the characters to its index. Then we can skip the characters immediately when we found a repeated character. Start with the non repeated position.
*/
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> map(256,-1);
        int start = -1;
        int count = 0;
        for(int i=0; i<s.size(); i++){
            if(map[s[i]]>start) start = map[s[i]];
            map[s[i]] = i;
            count = max(count, i-start);
        }
        return count;
    }
};

File: IndexGame/3longestSubstringWithoutRepeatingCharacters.py
#index

#save the index if it not appears in dic
#if appears later, i+1

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = nums = 0
        dic = {}
        for i in xrange(len(s)):
            if s[i] in dic and start <= dic[s[i]]:
                start = dic[s[i]]+1
            else:
                nums = max(nums,i-start+1)
            dic[s[i]] = i
        return nums